### PR TITLE
Suggest adding a `=>` after match guard

### LINKED ISSUE #78585: Confusing parse error for missing `=>` after match guard
Can you spot the mistake?

```rust
fn main() {
    let x = 1;
    let y = 2;
    let value = 3;

    match value {
        Some(x) if x == y {
            self.next_token()?;
            Ok(true)
        },
        _ => {
            Ok(false)
        }
    }
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cd6c0119180a26e53a07cda6cf61adcc))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error: expected identifier, found keyword `self`
 --> src/main.rs:8:13
  |
7 |         Some(x) if x == y {
  |                         - while parsing this struct
8 |             self.next_token()?;
  |             ^^^^ expected identifier, found keyword

error: expected one of `.`, `=>`, `?`, or an operator, found `,`
  --> src/main.rs:10:10
   |
10 |         },
   |          ^ expected one of `.`, `=>`, `?`, or an operator

error: aborting due to 2 previous errors

error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

### PR DESCRIPTION
**Pull request by TaKO8Ki**
closes #78585

### CODE REVIEW

**Comment by fmease** [comment]
Does the following still parse after you change?

```rust
fn main() {
    let x = It { a: 1 };
    match x {
        Some(x) if x == It { a: 0 } => {}
        _ => {}
    }
}
struct It { a: i32 }
```

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -2617,6 +2626,26 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_match_guard_cond(&mut self) -> PResult<'a, P<Expr>> {
+        let snapshot = self.clone();
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/expr.rs:2630]: The only issue I'm seeing here is that this pessimises the parsing of *every* match arm. In this case I think it is reasonable to instead catch the problem in line 2499 and only in the error arm try to conditionally parse the match arm body. Probably also do that for the substitution cases handled right below.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -2617,6 +2626,26 @@ impl<'a> Parser<'a> {
...
+        let mut cond = self.parse_expr()?;
+        let (is_lhs_struct, is_rhs_struct) = match cond.kind {
+            ExprKind::Binary(_, ref lhs, ref rhs) => {
+                (matches!(lhs.kind, ExprKind::Struct(..)), matches!(rhs.kind, ExprKind::Struct(..)))
+            }
+            _ => (false, false),
+        };
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/expr.rs:2637]: As you recognize here, we might want to also detect when a struct literal was parsed that instead was supposed to be an ident and the match arm body, but that is not as critical, I think.

**Comment by Dylan-DPC** [comment]
@TaKO8Ki any updates on this?

**Comment by Dylan-DPC** [comment]
Closing this as inactive

**Pull request closed without merging**